File: ErmisClient/Desktop/src/github/koukobin/ermis/client/main/java/database/Message.java
/* Copyright (C) 2025 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package github.koukobin.ermis.client.main.java.database;

import java.util.Arrays;
import java.util.Objects;

import github.koukobin.ermis.common.message_types.ClientContentType;

/**
 * @author Ilias Koukovinis
 *
 */
public class Message {
    private String username;
    private int clientID;
    private int messageID;
    private int chatSessionID;
    private int chatSessionIndex;
    private byte[] text;
    private byte[] fileName;
    private byte[] imageBytes;
    private long timeWritten;
    private ClientContentType contentType;
    private boolean isSent;

    public Message(String username, int clientID, int messageID, int chatSessionID, int chatSessionIndex,
                   byte[] text, byte[] fileName, long timeWritten, ClientContentType contentType, boolean isSent) {
        this.username = username;
        this.clientID = clientID;
        this.messageID = messageID;
        this.chatSessionID = chatSessionID;
        this.chatSessionIndex = chatSessionIndex;
        this.text = text.clone();
        this.fileName = fileName.clone();
        this.timeWritten = timeWritten;
        this.contentType = contentType;
        this.isSent = isSent;
    }

    // Empty constructor with default values
    public Message() {
        this.username = "";
        this.clientID = 0;
        this.messageID = 0;
        this.chatSessionID = 0;
        this.chatSessionIndex = 0;
        this.text = null;
        this.fileName = null;
        this.timeWritten = 0;
        this.contentType = ClientContentType.TEXT; // Assuming default content type
        this.isSent = false;
    }

    public String getText() {
        return text != null ? new String(text) : "";
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public void setIsSent(boolean isSent) {
        this.isSent = isSent;
    }

    public void setClientID(int clientID) {
        this.clientID = clientID;
    }

    public void setMessageID(int messageID) {
        this.messageID = messageID;
    }

    public void setChatSessionID(int chatSessionID) {
        this.chatSessionID = chatSessionID;
    }

    public void setChatSessionIndex(int chatSessionIndex) {
        this.chatSessionIndex = chatSessionIndex;
    }

    public void setText(byte[] text) {
        this.text = text.clone();
    }

    public void setFileName(byte[] fileName) {
        this.fileName = fileName.clone();
    }
    
    public void setImageBytes(byte[] imageBytes) {
    	this.imageBytes = imageBytes.clone();
    }

    public void setTimeWritten(long timeWritten) {
        this.timeWritten = timeWritten;
    }

    public void setContentType(ClientContentType contentType) {
        this.contentType = contentType;
    }

    public String getUsername() {
        return username;
    }

    public int getClientID() {
        return clientID;
    }

    public int getMessageID() {
        return messageID;
    }

    public int getChatSessionID() {
        return chatSessionID;
    }

    public int getChatSessionIndex() {
        return chatSessionIndex;
    }

    public byte[] getFileName() {
        return fileName.clone();
    }

    public byte[] getImageBytes() {
        return imageBytes.clone();
    }
    
    public long getTimeWritten() {
        return timeWritten;
    }

    public ClientContentType getContentType() {
        return contentType;
    }

    public boolean isSent() {
        return isSent;
    }

    @Override
    public int hashCode() {
        return Objects.hash(messageID);
    }

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (obj == null || getClass() != obj.getClass()) {
			return false;
		}

		Message message = (Message) obj;
		return chatSessionID == message.chatSessionID &&
                chatSessionIndex == message.chatSessionIndex &&
                clientID == message.clientID &&
                contentType == message.contentType &&
                messageID == message.messageID &&
                Arrays.equals(text, message.text) &&
                Arrays.equals(fileName, message.fileName) &&
                timeWritten == message.timeWritten &&
                username.equals(message.username);
    }

    @Override
    public String toString() {
        return "Message{" +
                "username='" + username + '\'' +
                ", clientID=" + clientID +
                ", messageID=" + messageID +
                ", chatSessionID=" + chatSessionID +
                ", chatSessionIndex=" + chatSessionIndex +
                ", text=" + Arrays.toString(text) +
                ", fileName=" + Arrays.toString(fileName) +
                ", timeWritten=" + timeWritten +
                ", contentType=" + contentType +
                '}';
    }
}

File: ErmisClient/Mobile/ermis_client/lib/client/common/message.dart
/* Copyright (C) 2024 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

import 'dart:convert';
import 'dart:typed_data';

import 'message_types/content_type.dart';

class Message {
  String _username;
  int _clientID;
  int _messageID;
  int _chatSessionID;
  int _chatSessionIndex;
  Uint8List? _text;
  Uint8List? _fileName;
  Uint8List? _imageBytes;
  int _timeWritten;
  MessageContentType _contentType;
  bool _isSent = false;

  Message({
    required String username,
    required int clientID,
    required int messageID,
    required int chatSessionID,
    required int chatSessionIndex,
    Uint8List? text,
    Uint8List? fileName,
    required int timeWritten,
    required MessageContentType contentType,
    required bool isSent,
  })  : _isSent = isSent,
        _contentType = contentType,
        _timeWritten = timeWritten,
        _fileName = fileName,
        _clientID = clientID,
        _messageID = messageID,
        _chatSessionIndex = chatSessionIndex,
        _text = text,
        _chatSessionID = chatSessionID,
        _username = username;

  Message.empty()
      : _username = '',
        _clientID = 0,
        _messageID = 0,
        _chatSessionID = 0,
        _chatSessionIndex = 0,
        _text = null,
        _fileName = null,
        _timeWritten = 0,
        _contentType = MessageContentType.text, // Assuming a default value
        _isSent = false;

  void setUsername(String username) => _username = username;
  void setIsSent(bool isSent) => _isSent = isSent;
  void setClientID(int clientID) => _clientID = clientID;
  void setMessageID(int messageID) => _messageID = messageID;
  void setChatSessionID(int chatSessionID) => _chatSessionID = chatSessionID;
  void setChatSessionIndex(int chatSessionIndex) => _chatSessionIndex = chatSessionIndex;
  void setText(Uint8List? text) => _text = text;
  void setFileName(Uint8List? fileName) => _fileName = fileName;
  set imageBytes(Uint8List? imageBytes) => _imageBytes = imageBytes;
  void setTimeWritten(int timeWritten) => _timeWritten = timeWritten;
  void setContentType(MessageContentType contentType) =>
      _contentType = contentType;

  String get username => _username;
  int get clientID => _clientID;
  int get messageID => _messageID;
  int get chatSessionID => _chatSessionID;
  int get chatSessionIndex => _chatSessionIndex;
  String get text {
    if (_text == null) {
      return "";
    }

    return utf8.decode(_text!.toList(), allowMalformed: true);
  }

  String get fileName {
    if (_fileName == null) {
      return "";
    }

    return utf8.decode(_fileName!.toList(), allowMalformed: true);
  }

  Uint8List? get imageBytes => _imageBytes;
  int get timeWritten => _timeWritten;
  MessageContentType get contentType => _contentType;
  bool get isSent => _isSent;

  @override
  int get hashCode => _messageID.hashCode;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! Message) return false;
    return _chatSessionID == other._chatSessionID &&
        _chatSessionIndex == other._chatSessionIndex &&
        _clientID == other._clientID &&
        _contentType == other._contentType &&
        _messageID == other._messageID &&
        _text == other._text &&
        _fileName == other._fileName &&
        _timeWritten == other._timeWritten &&
        _username == other._username;
  }

  @override
  String toString() {
    return 'Message{username: $_username, clientID: $_clientID, messageID: $_messageID, chatSessionID: $_chatSessionID, '
        'chatSessionIndex $_chatSessionIndex, text: $_text, fileName: $_fileName, timeWritten: $_timeWritten, contentType: $_contentType}';
  }
}
